modules/sdk/src/main/scala/org/tessellation/sdk/domain/healthcheck/consensus/HealthCheckConsensusRound.scala

```diff
@@ -0,0 +1,122 @@
+package org.tessellation.sdk.domain.healthcheck.consensus
+
+import cats.Applicative
+import cats.effect.{Clock, Ref, Spawn}
+import cats.syntax.contravariantSemigroupal._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.sdk.domain.gossip.Gossip
+import org.tessellation.sdk.domain.healthcheck.consensus.types._
+
+class HealthCheckConsensusRound[F[_]: Spawn: Clock, A <: HealthCheckStatus, B <: ConsensusHealthStatus[A]](
+  key: HealthCheckKey,
+  roundId: HealthCheckRoundId,
+  driver: HealthCheckConsensusDriver[A, B],
+  startedAt: Long,
+  peers: Ref[F, Set[PeerId]],
+  roundIds: Ref[F, Set[HealthCheckRoundId]],
+  proposals: Ref[F, Map[PeerId, B]],
+  gossip: Gossip[F],
+  selfId: PeerId
+) {
+
+  def start: F[Unit] =
+    Spawn[F].start {
+      sendProposal
+    }.void
+
+  def getPeers: F[Set[PeerId]] = peers.get
+
+  def manageAbsent(absentPeers: Set[PeerId]): F[Unit] =
+    proposals.get
+      .map(_.keySet)
+      .map(absentPeers -- _)
+      .flatMap { toRemove =>
+        peers.modify { p =>
+          (p -- toRemove, ())
+        }
+      }
+
+  def isFinished: F[Boolean] = allProposalsReceived
+
+  def processProposal(proposal: B): F[Unit] =
+    if (proposal.key == key) {
+      proposals.modify { m =>
+        def updated =
+          m.get(proposal.owner)
+            .fold(m + (proposal.owner -> proposal))(_ => m)
+
+        (updated, ())
+      }.flatMap { _ =>
+        roundIds.modify { m =>
+          (m + proposal.roundId, ())
+        }
+      }
+    } else Applicative[F].unit
+
+  def getRoundIds: F[Set[HealthCheckRoundId]] = roundIds.get
+
+  def addParallelRounds(key: HealthCheckKey)(roundIds: Set[HealthCheckRoundId]): F[Unit] = ???
+
+  def calculateOutcome: F[HealthCheckConsensusDecision] = {
+    def status: A = ???
+
+    proposals.get.map(_.values.toList).map { received =>
+      driver.calculateConsensusOutcome(key, status, selfId, received)
+    }
+  }
+
+  def manage: F[Unit] =
+    Clock[F].monotonic.map(_.toNanos).flatMap { currentTime =>
+      sendProposal // Note: check elapsed time and execute additional tasks
+    }
+
+  def generateHistoricalData(decision: HealthCheckConsensusDecision): F[HistoricalRound] = ???
+
+  def ownConsensusHealthStatus: F[B] = {
+    def status: A = ???
+    Applicative[F].pure(driver.consensusHealthStatus(key, status, roundId, selfId))
+  }
+
+  private def sendProposal: F[Unit] =
+    ownConsensusHealthStatus.flatMap(gossip.spread)
+
+  private def allProposalsReceived: F[Boolean] =
+    (peers.get, proposals.get.map(_.keySet))
```

Comment:
I guess we also need the functionality for adding peers to `peers` Set, based on the incoming proposals. Though I'm not sure what algorithm/flow you intended for the final implementation. Currently we are only subtracting from the initial locally known peers - this may influence deterministicity for the global healthcheck outcome on the whole cluster - we need to stride towards determinsticity on the cluster level.

Reply:
It's not the initial list of peers, but the final peers that are required to make a proposal.
It starts with the initial list of peers but not necessary it is the same at the end.
We may want to add peers to `peers` later but I didn't see that in your code so I didn't want to change it for now

Reply:
Adding is there in constellation - hence idea of proxy peers. And yeah that's exactly what I meant, we start with initial peers and after that we only have functionality for removing peers with the `manageAbsent` arriving at final peers set. I think it may not be valid as I mentioned in the first comment.